`tests/test_run_app.py::TestShutdown` hang on py3.13 with `aiohappyeyeballs == 2.4.2`

### Describe the bug

After upgrading to `aiohappyeyeballs == 2.4.2`, the following four tests started hanging with Python 3.13.0rc2:

```
tests/test_run_app.py::TestShutdown::test_shutdown_new_conn_rejected
tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler
tests/test_run_app.py::TestShutdown::test_shutdown_timeout_not_reached
tests/test_run_app.py::TestShutdown::test_shutdown_wait_for_handler
```

After downgrading to `aiohappyeyeballs == 2.4.0`, all tests pass again.

FWICS CI hit it already in #9310, but I think you can't easily guess which tests hanged from its output.

### To Reproduce

1. Upgrade to `aiohappyeyeballs == 2.4.2`.
2. Run tests with Python 3.13.


### Expected behavior

Tests passing :-).

### Logs/tracebacks

```python-traceback
(this is after SIGTERM-ing hanging tests)

========================================================= test session starts =========================================================
platform linux -- Python 3.13.0rc2, pytest-8.3.3, pluggy-1.5.0 -- /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7-python3_13/install/usr/bin/python3.13
cachedir: .pytest_cache
rootdir: /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7
configfile: setup.cfg
collecting ... collected 8 items

tests/test_run_app.py::TestShutdown::test_shutdown_wait_for_handler FAILED                                                       [1/8]
tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler FAILED                                                        [2/8]
tests/test_run_app.py::TestShutdown::test_shutdown_timeout_not_reached FAILED                                                    [3/8]
tests/test_run_app.py::TestShutdown::test_shutdown_new_conn_rejected FAILED                                                      [4/8]
tests/test_run_app.py::TestShutdown::test_shutdown_pending_handler_responds PASSED                                               [5/8]
tests/test_run_app.py::TestShutdown::test_shutdown_close_idle_keepalive PASSED                                                   [6/8]
tests/test_run_app.py::TestShutdown::test_shutdown_close_websockets PASSED                                                       [7/8]
tests/test_run_app.py::TestShutdown::test_shutdown_handler_cancellation_suppressed PASSED                                        [8/8]

============================================================== FAILURES ===============================================================
_____________________________________________ TestShutdown.test_shutdown_wait_for_handler _____________________________________________

self = <test_run_app.TestShutdown object at 0x7f9d96c1c190>, aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>

    def test_shutdown_wait_for_handler(
        self, aiohttp_unused_port: Callable[[], int]
    ) -> None:
        port = aiohttp_unused_port()
        finished = False
    
        async def task():
            nonlocal finished
            await asyncio.sleep(2)
            finished = True
    
>       t, connection_count = self.run_app(port, 3, task)

aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>
finished   = True
port       = 38969
self       = <test_run_app.TestShutdown object at 0x7f9d96c1c190>
task       = <function TestShutdown.test_shutdown_wait_for_handler.<locals>.task at 0x7f9d968a7920>

tests/test_run_app.py:1007: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test_run_app.TestShutdown object at 0x7f9d96c1c190>, port = 38969, timeout = 3
task = <function TestShutdown.test_shutdown_wait_for_handler.<locals>.task at 0x7f9d968a7920>, extra_test = None

    def run_app(self, port: int, timeout: int, task, extra_test=None) -> asyncio.Task:
        num_connections = -1
    
        class DictRecordClear(dict):
            def clear(self):
                nonlocal num_connections
                # During Server.shutdown() we want to know how many connections still
                # remained before it got cleared. If the handler completed successfully
                # the connection should've been removed already. If not, this may
                # indicate a memory leak.
                num_connections = len(self)
                super().clear()
    
        class ServerWithRecordClear(web.Server):
            def __init__(self, *args, **kwargs):
                super().__init__(*args, **kwargs)
                self._connections = DictRecordClear()
    
        async def test() -> None:
            await asyncio.sleep(0.5)
            async with ClientSession() as sess:
                for _ in range(5):  # pragma: no cover
                    try:
                        with pytest.raises(asyncio.TimeoutError):
                            async with sess.get(
                                f"http://localhost:{port}/",
                                timeout=ClientTimeout(total=0.1),
                            ):
                                pass
                    except ClientConnectorError:
                        await asyncio.sleep(0.5)
                    else:
                        break
                async with sess.get(f"http://localhost:{port}/stop"):
                    pass
    
                if extra_test:
                    await extra_test(sess)
    
        async def run_test(app: web.Application) -> None:
            nonlocal test_task
            test_task = asyncio.create_task(test())
            yield
            await test_task
    
        async def handler(request: web.Request) -> web.Response:
            nonlocal t
            t = asyncio.create_task(task())
            await t
            return web.Response(text="FOO")
    
        t = test_task = None
        app = web.Application()
        app.cleanup_ctx.append(run_test)
        app.router.add_get("/", handler)
        app.router.add_get("/stop", self.stop)
    
        with mock.patch("aiohttp.web_app.Server", ServerWithRecordClear):
            web.run_app(app, port=port, shutdown_timeout=timeout)
>       assert test_task.exception() is None
E       asyncio.exceptions.CancelledError

DictRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.DictRecordClear'>
ServerWithRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.ServerWithRecordClear'>
app        = <Application 0x7f9d96c18ec0>
extra_test = None
handler    = <function TestShutdown.run_app.<locals>.handler at 0x7f9d968a7ce0>
num_connections = 0
port       = 38969
run_test   = <function TestShutdown.run_app.<locals>.run_test at 0x7f9d968a7c40>
self       = <test_run_app.TestShutdown object at 0x7f9d96c1c190>
t          = <Task finished name='Task-10' coro=<TestShutdown.test_shutdown_wait_for_handler.<locals>.task() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:1002> result=None>
task       = <function TestShutdown.test_shutdown_wait_for_handler.<locals>.task at 0x7f9d968a7920>
test       = <function TestShutdown.run_app.<locals>.test at 0x7f9d968a79c0>
test_task  = <Task cancelled name='Task-2' coro=<TestShutdown.run_app.<locals>.test() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:952>>
timeout    = 3

tests/test_run_app.py:993: CancelledError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
======== Running on http://0.0.0.0:38969 ========
(Press CTRL+C to quit)
_____________________________________________ TestShutdown.test_shutdown_timeout_handler ______________________________________________

self = <test_run_app.TestShutdown object at 0x7f9d96c1cb90>, aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>

    def test_shutdown_timeout_handler(
        self, aiohttp_unused_port: Callable[[], int]
    ) -> None:
        port = aiohttp_unused_port()
        finished = False
    
        async def task():
            nonlocal finished
            await asyncio.sleep(2)
            finished = True
    
>       t, connection_count = self.run_app(port, 1, task)

aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>
finished   = True
port       = 45001
self       = <test_run_app.TestShutdown object at 0x7f9d96c1cb90>
task       = <function TestShutdown.test_shutdown_timeout_handler.<locals>.task at 0x7f9d967e5940>

tests/test_run_app.py:1025: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test_run_app.TestShutdown object at 0x7f9d96c1cb90>, port = 45001, timeout = 1
task = <function TestShutdown.test_shutdown_timeout_handler.<locals>.task at 0x7f9d967e5940>, extra_test = None

    def run_app(self, port: int, timeout: int, task, extra_test=None) -> asyncio.Task:
        num_connections = -1
    
        class DictRecordClear(dict):
            def clear(self):
                nonlocal num_connections
                # During Server.shutdown() we want to know how many connections still
                # remained before it got cleared. If the handler completed successfully
                # the connection should've been removed already. If not, this may
                # indicate a memory leak.
                num_connections = len(self)
                super().clear()
    
        class ServerWithRecordClear(web.Server):
            def __init__(self, *args, **kwargs):
                super().__init__(*args, **kwargs)
                self._connections = DictRecordClear()
    
        async def test() -> None:
            await asyncio.sleep(0.5)
            async with ClientSession() as sess:
                for _ in range(5):  # pragma: no cover
                    try:
                        with pytest.raises(asyncio.TimeoutError):
                            async with sess.get(
                                f"http://localhost:{port}/",
                                timeout=ClientTimeout(total=0.1),
                            ):
                                pass
                    except ClientConnectorError:
                        await asyncio.sleep(0.5)
                    else:
                        break
                async with sess.get(f"http://localhost:{port}/stop"):
                    pass
    
                if extra_test:
                    await extra_test(sess)
    
        async def run_test(app: web.Application) -> None:
            nonlocal test_task
            test_task = asyncio.create_task(test())
            yield
            await test_task
    
        async def handler(request: web.Request) -> web.Response:
            nonlocal t
            t = asyncio.create_task(task())
            await t
            return web.Response(text="FOO")
    
        t = test_task = None
        app = web.Application()
        app.cleanup_ctx.append(run_test)
        app.router.add_get("/", handler)
        app.router.add_get("/stop", self.stop)
    
        with mock.patch("aiohttp.web_app.Server", ServerWithRecordClear):
            web.run_app(app, port=port, shutdown_timeout=timeout)
>       assert test_task.exception() is None
E       asyncio.exceptions.CancelledError

DictRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.DictRecordClear'>
ServerWithRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.ServerWithRecordClear'>
app        = <Application 0x7f9d96dbfed0>
extra_test = None
handler    = <function TestShutdown.run_app.<locals>.handler at 0x7f9d967e5440>
num_connections = 0
port       = 45001
run_test   = <function TestShutdown.run_app.<locals>.run_test at 0x7f9d967e5300>
self       = <test_run_app.TestShutdown object at 0x7f9d96c1cb90>
t          = <Task finished name='Task-25' coro=<TestShutdown.test_shutdown_timeout_handler.<locals>.task() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:1020> result=None>
task       = <function TestShutdown.test_shutdown_timeout_handler.<locals>.task at 0x7f9d967e5940>
test       = <function TestShutdown.run_app.<locals>.test at 0x7f9d967e51c0>
test_task  = <Task cancelled name='Task-17' coro=<TestShutdown.run_app.<locals>.test() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:952>>
timeout    = 1

tests/test_run_app.py:993: CancelledError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
======== Running on http://0.0.0.0:45001 ========
(Press CTRL+C to quit)
___________________________________________ TestShutdown.test_shutdown_timeout_not_reached ____________________________________________

self = <test_run_app.TestShutdown object at 0x7f9d96d3ba80>, aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>

    def test_shutdown_timeout_not_reached(
        self, aiohttp_unused_port: Callable[[], int]
    ) -> None:
        port = aiohttp_unused_port()
        finished = False
    
        async def task():
            nonlocal finished
            await asyncio.sleep(1)
            finished = True
    
        start_time = time.time()
>       t, connection_count = self.run_app(port, 15, task)

aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>
finished   = True
port       = 58897
self       = <test_run_app.TestShutdown object at 0x7f9d96d3ba80>
start_time = 1727516905.0695245
task       = <function TestShutdown.test_shutdown_timeout_not_reached.<locals>.task at 0x7f9d967e5760>

tests/test_run_app.py:1044: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test_run_app.TestShutdown object at 0x7f9d96d3ba80>, port = 58897, timeout = 15
task = <function TestShutdown.test_shutdown_timeout_not_reached.<locals>.task at 0x7f9d967e5760>, extra_test = None

    def run_app(self, port: int, timeout: int, task, extra_test=None) -> asyncio.Task:
        num_connections = -1
    
        class DictRecordClear(dict):
            def clear(self):
                nonlocal num_connections
                # During Server.shutdown() we want to know how many connections still
                # remained before it got cleared. If the handler completed successfully
                # the connection should've been removed already. If not, this may
                # indicate a memory leak.
                num_connections = len(self)
                super().clear()
    
        class ServerWithRecordClear(web.Server):
            def __init__(self, *args, **kwargs):
                super().__init__(*args, **kwargs)
                self._connections = DictRecordClear()
    
        async def test() -> None:
            await asyncio.sleep(0.5)
            async with ClientSession() as sess:
                for _ in range(5):  # pragma: no cover
                    try:
                        with pytest.raises(asyncio.TimeoutError):
                            async with sess.get(
                                f"http://localhost:{port}/",
                                timeout=ClientTimeout(total=0.1),
                            ):
                                pass
                    except ClientConnectorError:
                        await asyncio.sleep(0.5)
                    else:
                        break
                async with sess.get(f"http://localhost:{port}/stop"):
                    pass
    
                if extra_test:
                    await extra_test(sess)
    
        async def run_test(app: web.Application) -> None:
            nonlocal test_task
            test_task = asyncio.create_task(test())
            yield
            await test_task
    
        async def handler(request: web.Request) -> web.Response:
            nonlocal t
            t = asyncio.create_task(task())
            await t
            return web.Response(text="FOO")
    
        t = test_task = None
        app = web.Application()
        app.cleanup_ctx.append(run_test)
        app.router.add_get("/", handler)
        app.router.add_get("/stop", self.stop)
    
        with mock.patch("aiohttp.web_app.Server", ServerWithRecordClear):
            web.run_app(app, port=port, shutdown_timeout=timeout)
>       assert test_task.exception() is None
E       asyncio.exceptions.CancelledError

DictRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.DictRecordClear'>
ServerWithRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.ServerWithRecordClear'>
app        = <Application 0x7f9d96c1d6d0>
extra_test = None
handler    = <function TestShutdown.run_app.<locals>.handler at 0x7f9d967e5e40>
num_connections = 0
port       = 58897
run_test   = <function TestShutdown.run_app.<locals>.run_test at 0x7f9d967e5d00>
self       = <test_run_app.TestShutdown object at 0x7f9d96d3ba80>
t          = <Task finished name='Task-40' coro=<TestShutdown.test_shutdown_timeout_not_reached.<locals>.task() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:1038> result=None>
task       = <function TestShutdown.test_shutdown_timeout_not_reached.<locals>.task at 0x7f9d967e5760>
test       = <function TestShutdown.run_app.<locals>.test at 0x7f9d967e5bc0>
test_task  = <Task cancelled name='Task-32' coro=<TestShutdown.run_app.<locals>.test() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:952>>
timeout    = 15

tests/test_run_app.py:993: CancelledError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
======== Running on http://0.0.0.0:58897 ========
(Press CTRL+C to quit)
____________________________________________ TestShutdown.test_shutdown_new_conn_rejected _____________________________________________

self = <test_run_app.TestShutdown object at 0x7f9d96890e90>, aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>

    def test_shutdown_new_conn_rejected(
        self, aiohttp_unused_port: Callable[[], int]
    ) -> None:
        port = aiohttp_unused_port()
        finished = False
    
        async def task() -> None:
            nonlocal finished
            await asyncio.sleep(9)
            finished = True
    
        async def test(sess: ClientSession) -> None:
            # Ensure we are in the middle of shutdown (waiting for task()).
            await asyncio.sleep(1)
            with pytest.raises(ClientConnectorError):
                # Use a new session to try and open a new connection.
                async with ClientSession() as sess:
                    async with sess.get(f"http://localhost:{port}/"):
                        pass
            assert finished is False
    
>       t, connection_count = self.run_app(port, 10, task, test)

aiohttp_unused_port = <function unused_port at 0x7f9d9725d3a0>
finished   = True
port       = 48985
self       = <test_run_app.TestShutdown object at 0x7f9d96890e90>
task       = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.task at 0x7f9d967e5a80>
test       = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.test at 0x7f9d967e62a0>

tests/test_run_app.py:1073: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test_run_app.TestShutdown object at 0x7f9d96890e90>, port = 48985, timeout = 10
task = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.task at 0x7f9d967e5a80>
extra_test = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.test at 0x7f9d967e62a0>

    def run_app(self, port: int, timeout: int, task, extra_test=None) -> asyncio.Task:
        num_connections = -1
    
        class DictRecordClear(dict):
            def clear(self):
                nonlocal num_connections
                # During Server.shutdown() we want to know how many connections still
                # remained before it got cleared. If the handler completed successfully
                # the connection should've been removed already. If not, this may
                # indicate a memory leak.
                num_connections = len(self)
                super().clear()
    
        class ServerWithRecordClear(web.Server):
            def __init__(self, *args, **kwargs):
                super().__init__(*args, **kwargs)
                self._connections = DictRecordClear()
    
        async def test() -> None:
            await asyncio.sleep(0.5)
            async with ClientSession() as sess:
                for _ in range(5):  # pragma: no cover
                    try:
                        with pytest.raises(asyncio.TimeoutError):
                            async with sess.get(
                                f"http://localhost:{port}/",
                                timeout=ClientTimeout(total=0.1),
                            ):
                                pass
                    except ClientConnectorError:
                        await asyncio.sleep(0.5)
                    else:
                        break
                async with sess.get(f"http://localhost:{port}/stop"):
                    pass
    
                if extra_test:
                    await extra_test(sess)
    
        async def run_test(app: web.Application) -> None:
            nonlocal test_task
            test_task = asyncio.create_task(test())
            yield
            await test_task
    
        async def handler(request: web.Request) -> web.Response:
            nonlocal t
            t = asyncio.create_task(task())
            await t
            return web.Response(text="FOO")
    
        t = test_task = None
        app = web.Application()
        app.cleanup_ctx.append(run_test)
        app.router.add_get("/", handler)
        app.router.add_get("/stop", self.stop)
    
        with mock.patch("aiohttp.web_app.Server", ServerWithRecordClear):
            web.run_app(app, port=port, shutdown_timeout=timeout)
>       assert test_task.exception() is None
E       asyncio.exceptions.CancelledError

DictRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.DictRecordClear'>
ServerWithRecordClear = <class 'test_run_app.TestShutdown.run_app.<locals>.ServerWithRecordClear'>
app        = <Application 0x7f9d96891940>
extra_test = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.test at 0x7f9d967e62a0>
handler    = <function TestShutdown.run_app.<locals>.handler at 0x7f9d967e5120>
num_connections = 0
port       = 48985
run_test   = <function TestShutdown.run_app.<locals>.run_test at 0x7f9d967e6520>
self       = <test_run_app.TestShutdown object at 0x7f9d96890e90>
t          = <Task finished name='Task-55' coro=<TestShutdown.test_shutdown_new_conn_rejected.<locals>.task() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:1058> result=None>
task       = <function TestShutdown.test_shutdown_new_conn_rejected.<locals>.task at 0x7f9d967e5a80>
test       = <function TestShutdown.run_app.<locals>.test at 0x7f9d967e6480>
test_task  = <Task cancelled name='Task-47' coro=<TestShutdown.run_app.<locals>.test() done, defined at /tmp/portage/dev-python/aiohttp-3.10.7/work/aiohttp-3.10.7/tests/test_run_app.py:952>>
timeout    = 10

tests/test_run_app.py:993: CancelledError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
======== Running on http://0.0.0.0:48985 ========
(Press CTRL+C to quit)
======================================================== slowest 10 durations =========================================================
85.24s call     tests/test_run_app.py::TestShutdown::test_shutdown_wait_for_handler
9.51s call     tests/test_run_app.py::TestShutdown::test_shutdown_new_conn_rejected
6.62s call     tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler
4.37s call     tests/test_run_app.py::TestShutdown::test_shutdown_timeout_not_reached
4.01s call     tests/test_run_app.py::TestShutdown::test_shutdown_pending_handler_responds
2.41s call     tests/test_run_app.py::TestShutdown::test_shutdown_handler_cancellation_suppressed
1.01s call     tests/test_run_app.py::TestShutdown::test_shutdown_close_websockets
1.01s call     tests/test_run_app.py::TestShutdown::test_shutdown_close_idle_keepalive
0.00s teardown tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler
0.00s setup    tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler
======================================================= short test summary info =======================================================
FAILED tests/test_run_app.py::TestShutdown::test_shutdown_wait_for_handler - asyncio.exceptions.CancelledError
FAILED tests/test_run_app.py::TestShutdown::test_shutdown_timeout_handler - asyncio.exceptions.CancelledError
FAILED tests/test_run_app.py::TestShutdown::test_shutdown_timeout_not_reached - asyncio.exceptions.CancelledError
FAILED tests/test_run_app.py::TestShutdown::test_shutdown_new_conn_rejected - asyncio.exceptions.CancelledError
=============================================== 4 failed, 4 passed in 114.53s (0:01:54) ===============================================
```


### Python Version

```console
$ python3.13 --version
Python 3.13.0rc2
```


### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.10.7
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /usr/lib/python3.13/site-packages
Requires: aiohappyeyeballs, aiosignal, attrs, frozenlist, multidict, yarl
Required-by: aiohttp-cors, aiohttp-oauthlib, aiohttp_socks, aioresponses, vdirsyncer
```


### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.1.0
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/lib/python3.13/site-packages
Requires: 
Required-by: aiohttp, grpclib, yarl
```


### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.13.1
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /usr/lib/python3.13/site-packages
Requires: idna, multidict
Required-by: aiohttp, vcrpy
```


### OS

Gentoo Linux amd64

### Related component

Server

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

--- a/CHANGES/9326.bugfix.rst
+++ b/CHANGES/9326.bugfix.rst
@@ -1,0 +1,1 @@
+Fixed cancellation leaking upwards on timeout -- by :user:`bdraco`.

--- a/aiohttp/helpers.py
+++ b/aiohttp/helpers.py
@@ -692,6 +692,17 @@ class TimerNoop(BaseTimerContext):
         return
 
 
+if sys.version_info >= (3, 11):
+
+    def _uncancel_task(task: "asyncio.Task[object]") -> None:
+        task.uncancel()
+
+else:
+
+    def _uncancel_task(task: "asyncio.Task[object]") -> None:
+        pass
+
+
 class TimerContext(BaseTimerContext):
     """Low resolution timeout context manager"""
 
@@ -723,10 +734,16 @@ class TimerContext(BaseTimerContext):
         exc_val: Optional[BaseException],
         exc_tb: Optional[TracebackType],
     ) -> Optional[bool]:
+        enter_task: Optional[asyncio.Task[Any]] = None
         if self._tasks:
-            self._tasks.pop()  # type: ignore[unused-awaitable]
+            enter_task = self._tasks.pop()
 
         if exc_type is asyncio.CancelledError and self._cancelled:
+            assert enter_task is not None
+            # The timeout was hit, and the task was cancelled
+            # so we need to uncancel the last task that entered the context manager
+            # since the cancellation should not leak out of the context manager
+            _uncancel_task(enter_task)
             raise asyncio.TimeoutError from None
         return None
 

--- a/aiohttp/helpers.py
+++ b/aiohttp/helpers.py
@@ -733,7 +733,7 @@ class TimerContext(BaseTimerContext):
         if self._tasks:
             enter_task = self._tasks.pop()
 
-        if self._cancelled and exc_type is asyncio.CancelledError:
+        if exc_type is asyncio.CancelledError and self._cancelled:
             assert enter_task is not None
             # The timeout was hit, and the task was cancelled
             # so we need to uncancel the last task that entered the context manager

--- a/aiohttp/helpers.py
+++ b/aiohttp/helpers.py
@@ -699,6 +699,7 @@ class TimerContext(BaseTimerContext):
         self._loop = loop
         self._tasks: List[asyncio.Task[Any]] = []
         self._cancelled = False
+        self._cancelling = 0
 
     def assert_timeout(self) -> None:
         """Raise TimeoutError if timer has already been cancelled."""
@@ -707,9 +708,14 @@ class TimerContext(BaseTimerContext):
 
     def __enter__(self) -> BaseTimerContext:
         task = asyncio.current_task(loop=self._loop)
-
         if task is None:
             raise RuntimeError("Timeout context manager should be used inside a task")
+
+        if sys.version_info >= (3, 11):
+            # Remember if the task was already cancelling
+            # so when we __exit__ we can decide if we should
+            # raise asyncio.TimeoutError or let the cancellation propagate
+            self._cancelling = task.cancelling()
 
         if self._cancelled:
             raise asyncio.TimeoutError from None
@@ -727,14 +733,18 @@ class TimerContext(BaseTimerContext):
         if self._tasks:
             enter_task = self._tasks.pop()
 
-        if exc_type is asyncio.CancelledError and self._cancelled:
+        if self._cancelled and exc_type is asyncio.CancelledError:
             assert enter_task is not None
             # The timeout was hit, and the task was cancelled
             # so we need to uncancel the last task that entered the context manager
             # since the cancellation should not leak out of the context manager
             if sys.version_info >= (3, 11):
-                enter_task.uncancel()
-            raise asyncio.TimeoutError from None
+                # If the task was already cancelling don't raise
+                # asyncio.TimeoutError and instead return None
+                # to allow the cancellation to propagate
+                if enter_task.uncancel() > self._cancelling:
+                    return None
+            raise asyncio.TimeoutError from exc_val
         return None
 
     def timeout(self) -> None:

--- a/aiohttp/helpers.py
+++ b/aiohttp/helpers.py
@@ -692,17 +692,6 @@ class TimerNoop(BaseTimerContext):
         return
 
 
-if sys.version_info >= (3, 11):
-
-    def _uncancel_task(task: "asyncio.Task[object]") -> None:
-        task.uncancel()
-
-else:
-
-    def _uncancel_task(task: "asyncio.Task[object]") -> None:
-        pass
-
-
 class TimerContext(BaseTimerContext):
     """Low resolution timeout context manager"""
 
@@ -743,7 +732,8 @@ class TimerContext(BaseTimerContext):
             # The timeout was hit, and the task was cancelled
             # so we need to uncancel the last task that entered the context manager
             # since the cancellation should not leak out of the context manager
-            _uncancel_task(enter_task)
+            if sys.version_info >= (3, 11):
+                enter_task.uncancel()
             raise asyncio.TimeoutError from None
         return None
 

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -451,6 +451,7 @@ async def test_timer_context_timeout_does_swallow_cancellation() -> None:
     """Verify that the TimerContext does not swallow cancellation."""
     loop = asyncio.get_running_loop()
     current_task = asyncio.current_task()
+    assert current_task is not None
     ctx = helpers.TimerContext(loop)
 
     async def task_with_timeout() -> None:

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -444,6 +444,40 @@ async def test_timer_context_timeout_does_not_leak_upward() -> None:
     assert current_task.cancelling() == 0
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 11), reason="Python 3.11+ is required for .cancelling()"
+)
+async def test_timer_context_timeout_does_swallow_cancellation() -> None:
+    """Verify that the TimerContext does not swallow cancellation."""
+    loop = asyncio.get_running_loop()
+    current_task = asyncio.current_task()
+    ctx = helpers.TimerContext(loop)
+
+    async def task_with_timeout():
+        nonlocal ctx
+        current_task = asyncio.current_task()
+        assert current_task is not None
+        with pytest.raises(asyncio.TimeoutError):
+            with ctx:
+                assert current_task.cancelling() == 0
+                await asyncio.sleep(1)
+
+    task = asyncio.create_task(task_with_timeout())
+    await asyncio.sleep(0)
+    task.cancel()
+    assert task.cancelling() == 1
+    ctx.timeout()
+
+    # Cancellation should not leak into the current task
+    assert current_task.cancelling() == 0
+    # Cancellation should not be swallowed if the task is cancelled
+    # and it also times out
+    await asyncio.sleep(0)
+    with pytest.raises(asyncio.CancelledError):
+        await task
+    assert task.cancelling() == 1
+
+
 def test_timer_context_no_task(loop: asyncio.AbstractEventLoop) -> None:
     with pytest.raises(RuntimeError):
         with helpers.TimerContext(loop):

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -425,6 +425,24 @@ def test_timer_context_not_cancelled() -> None:
         assert not m_asyncio.current_task.return_value.cancel.called
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 11), reason="Python 3.11+ is required for .cancelling()"
+)
+async def test_timer_context_timeout_does_not_leak_upward() -> None:
+    """Verify that the TimerContext does not leak cancellation outside the context manager."""
+    loop = asyncio.get_running_loop()
+    ctx = helpers.TimerContext(loop)
+    current_task = asyncio.current_task()
+    with pytest.raises(asyncio.TimeoutError):
+        with ctx:
+            assert current_task.cancelling() == 0
+            loop.call_soon(ctx.timeout)
+            await asyncio.sleep(1)
+
+    # After the context manager exits, the task should no longer be cancelling
+    assert current_task.cancelling() == 0
+
+
 def test_timer_context_no_task(loop: asyncio.AbstractEventLoop) -> None:
     with pytest.raises(RuntimeError):
         with helpers.TimerContext(loop):

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -433,6 +433,7 @@ async def test_timer_context_timeout_does_not_leak_upward() -> None:
     loop = asyncio.get_running_loop()
     ctx = helpers.TimerContext(loop)
     current_task = asyncio.current_task()
+    assert current_task is not None
     with pytest.raises(asyncio.TimeoutError):
         with ctx:
             assert current_task.cancelling() == 0

--- a/tests/test_helpers.py
+++ b/tests/test_helpers.py
@@ -453,13 +453,13 @@ async def test_timer_context_timeout_does_swallow_cancellation() -> None:
     current_task = asyncio.current_task()
     ctx = helpers.TimerContext(loop)
 
-    async def task_with_timeout():
+    async def task_with_timeout() -> None:
         nonlocal ctx
-        current_task = asyncio.current_task()
-        assert current_task is not None
+        new_task = asyncio.current_task()
+        assert new_task is not None
         with pytest.raises(asyncio.TimeoutError):
             with ctx:
-                assert current_task.cancelling() == 0
+                assert new_task.cancelling() == 0
                 await asyncio.sleep(1)
 
     task = asyncio.create_task(task_with_timeout())

